### ISSUE TITLE
Schema causes infinite loop

### ISSUE THREAD

**Issue by chrismcg** [jonasschmidt/ex_json_schema#11]
The following causes an infinite loop on ex_json_schema 0.5.0 but not on the json-schema ruby gem.

```
{
  "$schema": "http://json-schema.org/draft-04/schema#",

  "oneOf": [
    { "$ref": "#definitions/invalid" },
    { "$ref": "#definitions/valid" }
  ],

  "type": "object",

  "definitions": {
    "invalid": {
      "additionalProperties": false,
      "properties": {
        "errors": {
          "type": "array",
          "items": {
          }
        }
      },
      "required": [
        "errors"
      ]
    },

    "valid": {
      "additionalProperties": false,
      "properties": {
        "results": {
          "type": "array",
          "items": {
            "type": "object",
            "required": ["data"],
            "properties": {
              "data": {
                "type": "object",
                "required": ["value"]
              }
            }
          }
        }
      }
    }
  }
}
```

Adding `/`s in the `oneOf` definition fixes this:

```
  "oneOf": [
    { "$ref": "#/definitions/invalid" },
    { "$ref": "#/definitions/valid" }
  ],
```

I'm not sure if this lib or the ruby version is incorrect but I thought I'd better report it.

**Comment by jonasschmidt** [comment]
Hmm, ok. I will have a look what the spec has to say about it. But if it's not valid, it should give back a schema validation error, not run into an infinite recursion. So either way, it has to be fixed.

**Comment by jonasschmidt** [comment]
It's a bit hard to tell since the whole referencing referencing behaviour is not defined in an unambiguous way in the spec (see also https://github.com/json-schema/json-schema/issues/83 for a lengthy discussion around optional inline dereferencing). But in your case, the `$ref` values are JSON Pointers (https://tools.ietf.org/html/rfc6901) and should begin with a `/`.

It might be that some validators are more lenient about it, but requiring the reference to be a valid JSON Pointer also removes some ambiguity around inline referencing, where `#definitions` could point to a subschema that is defined inline.

In any case, `#/definitions/...` is the correct way to do it and I will try to turn this into a validation error instead of an infinite recursion :)

**Comment by jonasschmidt** [comment]
Alright, it's "fixed" in `0.5.1` to raise a `ExJsonSchema.Schema.InvalidSchemaError` when encountering invalid paths in `$ref`s.

Thanks for catching that one :)

**Comment by chrismcg** [comment]
:+1: thanks for fixing
